web/src/narrow.js

```diff
@@ -471,6 +471,16 @@ export function activate(raw_terms, opts) {
         }
         assert(msg_list !== undefined);
 
+        // Put the narrow terms in the URL fragment.
+        // Disabled when the URL fragment was the source
+        // of this narrow, but not if the fragment had
+        // a target message ID that has been moved.
+        // Important to call it before updating the current message list to avoid
+        // unexpected behaviour where current message list is different from hash.
+        if (opts.change_hash) {
+            save_narrow(terms, opts.trigger);
+        }
+
```

Comment:
Does it actually need to happen before `message_lists.update_current_message_list`, or just at the same time as it?

I wonder if we should be reworking things to have a single function that does both updates in a follow-up; `save_narrow` is such a weird function name...

Reply:
This change seems very plausibly correct, but I'd really like to understand the precise ordering problem that caused this to fix the back button, that we can make this comment as good as possible.

Reply:
I added this comment, but it likely needs further revision.

``` diff
diff --git a/web/src/narrow.js b/web/src/narrow.js
index 34741ee6e2..eeba23f186 100644
--- a/web/src/narrow.js
+++ b/web/src/narrow.js
@@ -466,12 +466,16 @@ export function activate(raw_terms, opts) {
         }
         assert(msg_list !== undefined);
 
-        // Put the narrow terms in the URL fragment.
-        // Disabled when the URL fragment was the source
-        // of this narrow, but not if the fragment had
-        // a target message ID that has been moved.
-        // Important to call it before updating the current message list to avoid
-        // unexpected behaviour where current message list is different from hash.
+        // Put the narrow terms in the URL fragment/hash.
+        //
+        // opts.change_hash will be false when the URL fragment was
+        // the source of this narrow, and the fragment was not a link to
+        // a specific target message ID that has been moved.
+        //
+        // This needs to be called at the same time as updating the
+        // current message list so that we don't need to think about
+        // bugs related to the URL fragment/hash being desynced from
+        // mesasge_lists.current.
         if (opts.change_hash) {
             save_narrow(terms, opts.trigger);
         }
```

Reply:
Pushed in #30060 to add a minor comment and rename `save_narrow`.